src/crimson/os/cyanstore/cyan_store.h

```diff
@@ -17,138 +17,277 @@
 #include "include/uuid.h"
 
 #include "crimson/os/cyanstore/cyan_object.h"
+#include "crimson/os/cyanstore/cyan_collection.h"
 #include "crimson/os/futurized_store.h"
 
 namespace ceph::os {
 class Transaction;
 }
 
 namespace crimson::os {
-class Collection;
 
 class CyanStore final : public FuturizedStore {
   const std::string path;
-  std::unordered_map<coll_t, boost::intrusive_ptr<Collection>> coll_map;
-  std::map<coll_t, boost::intrusive_ptr<Collection>> new_coll_map;
   uint64_t used_bytes = 0;
   uuid_d osd_fsid;
 
+  class ShardStores {
+    CyanStore *cyan_store;
+    const std::string path;
+    std::unordered_map<coll_t, boost::intrusive_ptr<Collection>> coll_map;
+    std::map<coll_t, boost::intrusive_ptr<Collection>> new_coll_map;
```

Comment:
These maps are per-shard, right?  I don't see a mechanism to ensure that the osd actually places the pg on the correct shard.  You wouldn't actually see the problem without starting an osd, stopping it, and then restarting it.

I think we actually want to address this rather than modify CyanStore to tolerate cross-core access.  Modify the list_collections interface to return a core_id_t for each Collection with NULL_CORE indicating no core preference.  AlienStore and SeaStore should return NULL_CORE for each collection, CyanStore here should return the correct core.

Modify PGShardMapping::maybe_create_pg to take an optional core parameter with NULL_CORE preserving the existing round-robin behavior:

```
  /**
   * maybe_create_pg
   *
   * Returns mapping for pgid, creates new one if it doesn't already exist.
   * If core != NULL_CORE, places pgid on core
   */
  core_id_t maybe_create_pg(spg_t pgid, core_id_t core = NULL_CORE) {
```

PGShardManager::load_pgs() should then pass the core it gets for each collection into maybe_create_pg.

Once SeaStore supports multicore, it'll return a real core as well.

Reply:
Got it,  the round-robin algorithm can't guarantee the pg create on the same core after restart.  will do it soon.